apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1092 @@
+as building the same variant with a different label would cause the dependency
+to stop applying.
+
+The discussion around environment markers surfaced more interesting problems.
+For a start, it was pointed out that there could be two different types
+of markers: markers matching the properties of the installed wheel (as outlined
+previously), and markers matching the current system state. For example,
+the latter could be used to install the `nvidia` plugin if CUDA is available without
+actually having to publish a separate CUDA variant of the main package (think
+JAX). However, this is a recent development and it has not been pursued yet.
+
+It was also pointed out that technically these kind of dependency specifiers can create
+potential conflicts. For example, since technically a wheel can have multiple
+values for a property, the code in listing 8. could end up pulling in two or three versions
+of `nvidia-cuda-runtime` simultaneously. We are still working on some
+of the finer implications of that, such as how it affects the creation
+of universal lockfiles, that attempt to account for all valid environment
+marker combinations.
+
+## Let's talk about security
+
+The most important part of the design that we're still working on is security.
+The installer, including all its (possibly vendored) dependencies, is sometimes
+found operating with elevated privileges, for example when it is used to install
+packages for all users, on a multi-user system. Variant provider plugins will
+be operating with the same level of privileges, and this has significant
+implications on how they are developed.
+
+<div style={{ textAlign: "center" }}>
+<figure style={{width: 'auto', margin: '0 2em', display: 'inline-block', verticalAlign: 'top'}}>
+  <img src="/posts/python-wheels-from-tags-to-variants/security-scopes.png"
+    width="906" height="458" alt="Change in scopes of privileges due to variant
+plugins. Prior to their support, only installer (and its dependencies) would be
+executed at package install (possibly with elevated privileges), and wheel
+contents would only be executed when the package/application is used.
+After the change, also variant plugin (and its dependencies) is executed
+at package install, which means executing contents of the downloaded wheels." />
+  <figcaption>Fig. 10. Scopes of privileges when executing code</figcaption>
+</figure>
+</div>
+
+There are two main risks being considered here. First, malicious code could
+be introduced into an existing variant provider or one of its dependencies;
+this is a similar risk to how such code could be introduced in the installer
+itself. Second, a new provider could be created that includes malicious code
+by design, and existing packages could start using it.
+
+It can be noted that these risks are similar to Python package build backends,
+that are also downloaded dynamically during the install process and that execute
```

Comment:
```suggestion
It can be noted that these risks are similar to those for build backends,
which are also downloaded dynamically during the install process and execute
```